specs/_features/epbs/beacon-chain.md

```diff
@@ -0,0 +1,627 @@
+        proposer_penalty_numerator = 0
+        for index in indexed_payload_attestation.attesting_indices:
+            for flag_index, weight in enumerate(PARTICIPATION_FLAG_WEIGHTS):
+                if has_flag(epoch_participation[index], flag_index):
+                    epoch_participation[index] = remove_flag(epoch_participation[index], flag_index)
+                    proposer_penalty_numerator += get_base_reward(state, index) * weight
+        # Penalize the proposer
+        proposer_penalty = Gwei(2*proposer_penalty_numerator // proposer_reward_denominator)
+        decrease_balance(state, proposer_index, proposer_penalty)
+        return
+
+    # Reward the proposer and set all the participation flags in case of correct attestations
+    proposer_reward_numerator = 0
+    for index in indexed_payload_attestation.attesting_indices:
+        for flag_index, weight in enumerate(PARTICIPATION_FLAG_WEIGHTS):
+            if not has_flag(epoch_participation[index], flag_index):
+                epoch_participation[index] = add_flag(epoch_participation[index], flag_index)
+                proposer_reward_numerator += get_base_reward(state, index) * weight
+
+    # Reward proposer
+    proposer_reward = Gwei(proposer_reward_numerator // proposer_reward_denominator)
+    increase_balance(state, proposer_index, proposer_reward)
+```
+
+#### New `verify_execution_payload_envelope_signature`
+
+```python
+def verify_execution_payload_envelope_signature(state: BeaconState, signed_envelope: SignedExecutionPayloadEnvelope) -> bool:
+    builder = state.validators[signed_envelope.message.builder_index]
+    signing_root = compute_signing_root(signed_envelope.message, get_domain(state, DOMAIN_BEACON_BUILDER))
+    return bls.Verify(builder.pubkey, signing_root, signed_envelope.signature)
+```
+
+#### Modified `process_execution_payload`
+*Note*: `process_execution_payload` is now an independent check in state transition. It is called when importing a signed execution payload proposed by the builder of the current slot.
+
+```python
+def process_execution_payload(state: BeaconState, signed_envelope: SignedExecutionPayloadEnvelope, execution_engine: ExecutionEngine, verify = True) -> None:
+    # Verify signature
+    if verify: 
+        assert verify_execution_payload_envelope_signature(state, signed_envelope)
+    envelope = signed_envelope.message
+    payload = envelope.payload
+    # Verify consistency with the beacon block
+    assert envelope.beacon_block_root == hash_tree_root(state.latest_block_header)
+
+    # Verify the withdrawals root
+    assert hash_tree_root(payload.withdrawals) == state.last_withdrawals_root
+
+    # Verify consistency with the committed header
+    committed_header = state.execution_payload_header
+    assert envelope.builder_index == committed_header.builder_index
+    assert committed_header.blob_kzg_commitments_root == hash_tree_root(envelope.blob_kzg_commitments)
+
+    if not envelope.payload_withheld: 
```

Comment:
In case the payload is made optional in the envelope, this logic should change accordingly